controllers/serviceexport/controller.go

```diff
@@ -0,0 +1,143 @@
+)
+
+// ServiceExport GVR
+var svcExportGVR schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    svcExportGroup,
+	Version:  svcExportVersion,
+	Resource: svcExportResource,
+}
+
+// ServiceImport GVR
+var svcImportGVR schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    svcImportGroup,
+	Version:  svcImportVersion,
+	Resource: svcImportResource,
+}
+
+// Controller (ServiceExport controller) runs on a member cluster and syncs Services
+// (via ServiceExport and ServiceImport objects) and endpointSlices (via EndpointSliceExport objects)
+// with the hub cluster.
+type Controller struct {
+	// Unique ID of the member cluster in the fleet
+	memberClusterID string
+	// Clients for operating on the built-in and dynamic resources in the member cluster
+	memberKubeClient    clientset.Interface
+	memberDynamicClient dynamic.Interface
+	// Client for operating on the dynamic resources in the hub cluster
+	hubDynamicClient dynamic.Interface
+	// Informers for ServiceExport, Service, and EndpointSlice resources in the member cluster
+	memberSvcExportInformer     informers.GenericInformer
+	memberSvcInformer           coreinformers.ServiceInformer
+	memberEndpointSliceInformer discoveryinformers.EndpointSliceInformer
+	// Informer for ServiceImport resources in the hub cluster
+	hubSvcImportInformer informers.GenericInformer
+	// Workqueues for syncing Services with the hub cluster
+	memberSvcSyncWorkqueue workqueue.RateLimitingInterface
+	hubSvcSyncWorkqueue    workqueue.RateLimitingInterface
+	// Workqueue for syncing EndpointSlices with the hub cluster
+	memberEndpointSliceSyncWorkqueue workqueue.RateLimitingInterface
+	// Trackers that track if a resource has been successfully sync'd before and its last sync'd spec
+	svcSyncStatusTracker           SvcSyncStatusTracker
+	endpointSliceSyncStatusTracker EndpointSliceSyncStatusTracker
+	// Broadcaster and recorder for handling events
+	eventBroadcaster record.EventBroadcaster
+	eventRecorder    record.EventRecorder
+}
+
+func New(
+	memberClusterID string,
+	memberKubeClient clientset.Interface,
+	memberDynamicClient dynamic.Interface,
+	hubKubeClient clientset.Interface,
+	hubDynamicClient dynamic.Interface,
+	memberSharedInformerFactory informers.SharedInformerFactory,
+	memberDynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+	hubDynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+) (*Controller, error) {
+	// Initialize informers
+	memberSvcExportInformer := memberDynamicInformerFactory.ForResource(svcExportGVR)
+	memberSvcInformer := memberSharedInformerFactory.Core().V1().Services()
+	memberEndpointSliceInformer := memberSharedInformerFactory.Discovery().V1().EndpointSlices()
+	hubSvcImportInformer := hubDynamicInformerFactory.ForResource(svcImportGVR)
+
+	// Initialize workqueues
+	memberSvcSyncWorkqueue := workqueue.NewNamedRateLimitingQueue(
+		workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay),
```

Comment:
should we define “workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay)” as a separate variable? it's used in multiple places.

Reply:
Unfortunately each workqueue needs its own `ItemExponentialFailureRateLimiter`, as it tracks queue specific rate limiting metrics.;Zhiying, do you mean here that for clarity reasons we should add a new line to instantiate a rate limiter for each workqueue declaration statement?

Reply:
oh, got it. Usually we would like to put the func call in one line for the better readability.

To shorten the line, we will create a variable.

eg,

memberSvcSyncWorkqueue := workqueue.NewNamedRateLimitingQueue(rateLimiter, "memberService"